python/mxnet/optimizer.py

```diff
@@ -670,9 +670,10 @@ def __init__(self, optimizer):
 
     def __call__(self, index, grad, weight):
         """Update weight given gradient and index"""
-        if index not in self.states:
-            self.states[index] = self.optimizer.create_state(index, weight)
-        self.optimizer.update(index, weight, grad, self.states[index])
+        key = (index, str(weight.context))
```

Comment:
shouldn't the index has already has device identity? e.g.

https://github.com/dmlc/mxnet/blob/master/python/mxnet/model.py#L116